pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,246 @@
+	if err != nil {
+		return nil, err
+	}
+
+	spec, err := container.Spec(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("getting spec for container %q: %w", containerID, err)
+	}
+
+	mountData := make([]runtimeclient.ContainerMountData, len(spec.Mounts))
+	for i := range spec.Mounts {
+		mount := spec.Mounts[i]
+		mountData[i] = runtimeclient.ContainerMountData{
+			Source:      mount.Source,
+			Destination: mount.Destination,
+		}
+	}
+
+	return &runtimeclient.ContainerDetailsData{
+		ContainerData: *containerData,
+		Pid:           int(task.Pid()),
+		CgroupsPath:   spec.Linux.CgroupsPath,
+		Mounts:        mountData,
 	}, nil
 }
+
+func (c *ContainerdClient) getContainerDataAndContainerAndTask(containerID string) (*runtimeclient.ContainerData, containerd.Container, containerd.Task, error) {
+	container, err := c.getContainerdContainer(containerID)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	task, err := getContainerdTask(container)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	containerData, err := taskAndContainerToContainerData(task, container)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return containerData, container, task, nil
+}
+
+// getContainerdContainer returns the corresponding container.Container instance to
+// the given id
+func (c *ContainerdClient) getContainerdContainer(id string) (containerd.Container, error) {
+	container, err := c.client.LoadContainer(context.TODO(), id)
+	if err != nil {
+		return nil, fmt.Errorf("loading container with id %q: %w", id, err)
+	}
+
+	if isSandboxContainer(container) {
+		log.Debugf("ContainerdClient: container %q is a sandbox container. Temporary skipping it", container.ID())
+		return nil, runtimeclient.ErrPauseContainer
+	}
+
+	return container, nil
+}
+
+// getContainerdTask returns the corresponding container.Task instance to
+// the given containerd.Container. If there is no task, nil is returned without an error
+func getContainerdTask(container containerd.Container) (containerd.Task, error) {
+	task, err := container.Task(context.TODO(), nil)
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return nil, fmt.Errorf("getting task for container %q: %w", container.ID(), err)
+		}
+		// else it is created, but not running
+		log.Debugf("No task for %q", container.ID())
+	}
+	return task, nil
+}
+
+// Constructs a ContainerData from a containerd.Task and containerd.Container
+// The extra containerd.Container parameter saves an additional call to the API
+func taskAndContainerToContainerData(task containerd.Task, container containerd.Container) (*runtimeclient.ContainerData, error) {
+	labels, err := container.Labels(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("listing labels of container %q: %w", container.ID(), err)
+	}
+
+	// If the container exists, but there is no running task/proc, the state is created
+	status := runtimeclient.StateCreated
+	if task != nil {
```

Comment:
I think this logic should be managed inside `getContainerdTask`. What about something like this?

```go
// containerTask represents the task information for a given container.
type containerTask struct {
	status string
	pid    uint32
}

// getContainerTask returns the containerTask information for a given container.
// If the container is not running yet, it returns a containerTask with status
// StateCreated and pid 0.
func getContainerTask(container containerd.Container) (*containerTask, error) {
	task, err := container.Task(context.TODO(), nil)
	if err != nil {
		// According to nerdctl, if there is no task, we can assume the
		// container was just created but it is not running yet:
		// https://github.com/containerd/nerdctl/blob/b0a75d880ef9e6d1f1a8752804c9087ee2d02f73/pkg/formatter/formatter.go#L48-L55
		if !errdefs.IsNotFound(err) {
			return nil, fmt.Errorf("getting task for container %q: %w", container.ID(), err)
		}

		t := &containerTask{
			status: runtimeclient.StateCreated,
			pid:    0,
		}
		log.Debugf("ContainerdClient: No task for %q. Assuming it is in %q status",
			container.ID(), t.status)

		return t, nil
	}

	containerdStatus, err := task.Status(context.TODO())
	if err != nil {
		return nil, fmt.Errorf("getting status of task for container %q: %w", container.ID(), err)
	}

	return &containerTask{
		status: processStatusStateToRuntimeClientState(containerdStatus.Status),
		pid:    task.Pid(),
	}, nil
}
```